TransformDialect::checkImplementsTransformOpInterface eliminated in Release builds, causing link failures

## Summary

`mlir::transform::detail::checkImplementsTransformOpInterface()` is eliminated by both clang and GCC at `-O2`/`-O3` when assertions are disabled (`-DNDEBUG`), because its entire function body consists of `assert()` calls and an `if` block guarding another `assert()`. The compiler correctly determines the function has no observable side effects and removes it.

However, this function is called from the template `TransformDialect::addOperationIfNotRegistered<OpTy>()` in the public header `TransformDialect.h`. Downstream projects that instantiate this template (e.g., mlir-aie) get an undefined reference at link time.

## Reproduction

Compile `mlir/lib/Dialect/Transform/IR/TransformDialect.cpp` with and without `-DNDEBUG`:

```bash
# Symbol present (assertions ON):
clang++ [flags] -O3 -std=c++17 -fvisibility=hidden -ffunction-sections \
  -o with_assert.o -c TransformDialect.cpp
nm -C with_assert.o | grep checkImplementsTransformOpInterface
# 0000000000000000 T mlir::transform::detail::checkImplementsTransformOpInterface(...)

# Symbol absent (assertions OFF):
clang++ [flags] -O3 -DNDEBUG -std=c++17 -fvisibility=hidden -ffunction-sections \
  -o no_assert.o -c TransformDialect.cpp
nm -C no_assert.o | grep checkImplementsTransformOpInterface
# (empty -- symbol eliminated)
```

| Compiler | Opt Level | NDEBUG | Symbol? |
|----------|-----------|--------|---------|
| clang 20 | -O3 | no | present |
| clang 20 | -O3 | yes | **absent** |
| clang 20 | -O2 | yes | **absent** |
| GCC 15 | -O3 | yes | **absent** |

## Link error

```
DialectExtension.cpp:(.text._ZN4mlir9transform16TransformDialect27addOperationIfNotRegisteredINS0_22VectorizeContractionOpEEEvv[...]+0xa0):
  undefined reference to `mlir::transform::detail::checkImplementsTransformOpInterface(llvm::StringRef, mlir::MLIRContext*)'
```

## Root cause

[`TransformDialect.cpp` lines 28-50](https://github.com/llvm/llvm-project/blob/979132a02d146ec79e2f046e31877516d7f32d20/mlir/lib/Dialect/Transform/IR/TransformDialect.cpp#L28-L50):

```cpp
void transform::detail::checkImplementsTransformOpInterface(
    StringRef name, MLIRContext *context) {
  RegisteredOperationName opName =
      *RegisteredOperationName::lookup(name, context);
  assert(/* ... */);          // <-- eliminated by NDEBUG
  if (!opName.hasInterface<PatternDescriptorOpInterface>() &&
      !opName.hasInterface<ConversionPatternDescriptorOpInterface>() &&
      !opName.hasInterface<TypeConverterBuilderOpInterface>()) {
    assert(/* ... */);        // <-- eliminated by NDEBUG
  }
}
```

With `-DNDEBUG`, `opName` is unused (only consumed by asserts), and the `if` block body is also an assert. The compiler eliminates the entire function.

The caller in [`TransformDialect.h`](https://github.com/llvm/llvm-project/blob/979132a02d146ec79e2f046e31877516d7f32d20/mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h):

```cpp
template <typename OpTy>
void TransformDialect::addOperationIfNotRegistered() {
  StringRef name = OpTy::getOperationName();
  // ...
  detail::checkImplementsTransformOpInterface(name, getContext());
}
```

This template is instantiated by downstream projects, generating a call to a symbol that no longer exists.

## Suggested fix

Give the function a non-assert side effect so it survives optimization. The `RegisteredOperationName::lookup` + dereference is valuable as a runtime check even in Release mode:

```cpp
void transform::detail::checkImplementsTransformOpInterface(
    StringRef name, MLIRContext *context) {
  RegisteredOperationName opName =
      *RegisteredOperationName::lookup(name, context);
  (void)opName;  // Ensure symbol survives in Release builds
  assert(/* ... existing asserts ... */);
}
```

Or alternatively, use `LLVM_ATTRIBUTE_USED` / `__attribute__((used))` on the function definition.

## Environment

- LLVM commit: `979132a02d146ec79e2f046e31877516d7f32d20` (Feb 2026)
- Also affects: any LLVM 23 checkout
- Compilers: clang 20.1.8, GCC 15 (Ubuntu 25.10)
- Downstream project: [Xilinx/mlir-aie](https://github.com/Xilinx/mlir-aie) (VectorizeContractionOp)

## Workaround

Build LLVM with `-DLLVM_ENABLE_ASSERTIONS=ON`. This keeps the function body intact.

--- a/mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h
+++ b/mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h
@@ -66,7 +66,7 @@ protected:
       : TransformDialectDataBase(TypeID::get<DerivedTy>(), ctx) {}
 };
 
-#ifndef NDEBUG
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
 namespace detail {
 /// Asserts that the operations provided as template arguments implement the
 /// TransformOpInterface and MemoryEffectsOpInterface. This must be a dynamic
@@ -79,7 +79,7 @@ void checkImplementsTransformOpInterface(StringRef name, MLIRContext *context);
 void checkImplementsTransformHandleTypeInterface(TypeID typeID,
                                                  MLIRContext *context);
 } // namespace detail
-#endif // NDEBUG
+#endif // LLVM_ENABLE_ABI_BREAKING_CHECKS
 } // namespace transform
 } // namespace mlir
 
@@ -256,10 +256,10 @@ void TransformDialect::addOperationIfNotRegistered() {
       RegisteredOperationName::lookup(TypeID::get<OpTy>(), getContext());
   if (!opName) {
     addOperations<OpTy>();
-#ifndef NDEBUG
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
     StringRef name = OpTy::getOperationName();
     detail::checkImplementsTransformOpInterface(name, getContext());
-#endif // NDEBUG
+#endif // LLVM_ENABLE_ABI_BREAKING_CHECKS
     return;
   }
 
@@ -289,10 +289,10 @@ void TransformDialect::addTypeIfNotRegistered() {
       });
   addTypes<Type>();
 
-#ifndef NDEBUG
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
   detail::checkImplementsTransformHandleTypeInterface(TypeID::get<Type>(),
                                                       getContext());
-#endif // NDEBUG
+#endif // LLVM_ENABLE_ABI_BREAKING_CHECKS
 }
 
 template <typename DataTy>

--- a/mlir/lib/Dialect/Transform/IR/TransformDialect.cpp
+++ b/mlir/lib/Dialect/Transform/IR/TransformDialect.cpp
@@ -24,7 +24,7 @@ using namespace mlir;
 #define GET_ATTRDEF_CLASSES
 #include "mlir/Dialect/Transform/IR/TransformAttrs.cpp.inc"
 
-#ifndef NDEBUG
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS
 void transform::detail::checkImplementsTransformOpInterface(
     StringRef name, MLIRContext *context) {
   // Since the operation is being inserted into the Transform dialect and the
@@ -61,7 +61,7 @@ void transform::detail::checkImplementsTransformHandleTypeInterface(
          "expected Transform dialect type to implement one of the three "
          "interfaces");
 }
-#endif // NDEBUG
+#endif // LLVM_ENABLE_ABI_BREAKING_CHECKS
 
 void transform::TransformDialect::initialize() {
   // Using the checked versions to enable the same assertions as for the ops

